deps: update rustls v0.20.1 -> v0.21.0

cpu opened the pull request:
## Description

This commit updates reqwest to use rustls 0.21.0, both as a direct dependency and through an update of tokio-rustls to 0.24.0, hyper-rustls to 0.24.0, quinn 0.10.0, and h3-quinn to 0.0.3.

One change is required in the reqwest codebase to adjust the import location of the `DigtallySignedStruct` type.

## Remaining work

- [X] wait for https://github.com/tokio-rs/tls/pull/137 to be merged
- [x] wait for tokio-rustls v0.24.0 to be published to crates.io
- [x] remove Cargo.toml patch for tokio-rustls
- [x] wait for https://github.com/rustls/hyper-rustls/pull/199 to be merged
- [x] wait for hyper-rustls v0.24.0 to be published to crates.io
- [x] remove Cargo.toml patch for hyper-rustls
- [x] wait for https://github.com/quinn-rs/quinn/pull/1515 to be merged
- [x] wait for quinn 0.10 to be published to crates.io https://github.com/quinn-rs/quinn/issues/1528
- [x] remove Cargo.toml patch for quinn
- [x] wait for https://github.com/hyperium/h3/pull/190 to be merged
- [x] wait for an h3 release tag.
- [x] remove Cargo.toml patch for h3
- [x] get PR approved

seanmonstar commented:
As a side note, part of my point of making h3 "experimental" is specifically so we can update rustls if need be, and "break" the h3 code for a time. At the same time, if the upgrade isn't urgent, it'd be kinder to keep the h3 stuff from being turned off.

seanmonstar commented:
FWIW, the h3-quinn crate was just able to update to quinn 0.9, at least.

marziply commented:
> Yay, looking forward to this as I want to migrate to rustls and need the ip address feature.

I'm in the same boat. The IP address feature is a blocker on a feature I am currently working on so I'm really keen to upgrade Reqwest when Rustls is bumped to v0.21.

cpu commented:
@seanmonstar The Quinn project cut a [0.10 release](https://github.com/quinn-rs/quinn/releases/tag/0.10.0) and I've opened a PR against h3 to update Quinn/Rustls there as well: https://github.com/hyperium/h3/pull/190

For the time being I've put a `patch` in to use my h3 PR branch. We can either wait for the h3 update to percolate out to merge this or I could drop the patch and we could merge without and break the experimental h3 support.

Ruben2424 commented:
I just published h3-quinn v0.0.3 with your changes.

cpu commented:
> I just published h3-quinn v0.0.3 with your changes.

Thanks!

I've removed the h3/h3-quinn Cargo patches and rebased. CI seems happy and the `cargo tree` is only showing Rustls 0.21.0 :tada: :lock: 

This branch should be ready for review now.

seanmonstar reviewed (approved):
The changes here are nice and clean, all the work was upstream. Thanks!

bartlomieju commented:
@seanmonstar can we expect a new `reqwest` release with this change soon? It's the last blocker for us in Deno to upgrade ourselves to rustls 0.21.0 and I've been waiting eagerly for this PR to land. Thanks for all the work!

seanmonstar commented:
yes, preparing in #1834

aviramha commented:
Yay!
